File: languages/go/03-advanced/inheritance-and-polymorphism/example/main.go
// Module focus: Treating different concrete types through one common interface.
// Why it matters: the example makes it possible to program against a shared behavioral
// abstraction before the learner tackles the exercises.

package main

import (
	"fmt"
	"math"
)

// Separate helpers keep the main path focused on how to program against a shared behavioral
// abstraction.
type Shape interface {
	Area() float64
	Name() string
}

type Rectangle struct {
	width  float64
	height float64
}

func (r Rectangle) Area() float64 {
	return r.width * r.height
}

func (r Rectangle) Name() string {
	return "Rectangle"
}

type Circle struct {
	radius float64
}

func (c Circle) Area() float64 {
	return math.Pi * c.radius * c.radius
}

func (c Circle) Name() string {
	return "Circle"
}

// Fixed inputs make the consequence of expecting inheritance syntax instead of
// composition/interfaces visible and repeatable.
func main() {
	// These values exercise the normal path before the exercises vary the documented boundaries.
	shapes := []Shape{
		Rectangle{width: 3.0, height: 4.0},
		Circle{radius: 2.0},
	}

	for _, shape := range shapes {
		// The printed result shows whether the program can use dynamic dispatch without unsafe type
		// assumptions.
		fmt.Printf("%s area: %.2f\n", shape.Name(), shape.Area())
	}
}

File: languages/typescript/03-advanced/inheritance-and-polymorphism/example/main.ts
// Module focus: Treating different concrete types through one common interface.
// Why it matters: the example makes it possible to program against a shared behavioral
// abstraction before the learner tackles the exercises.

abstract class Shape {
    // The base type defines the behavior the loop can depend on.
    abstract name(): string;
    abstract area(): number;
}

class Circle extends Shape {
    constructor(readonly radius: number) {
        super();
    }

    name(): string {
        return "Circle";
    }

    area(): number {
        return Math.PI * this.radius * this.radius;
    }
}

class Rectangle extends Shape {
    constructor(
        readonly width: number,
        readonly height: number,
    ) {
        super();
    }

    name(): string {
        return "Rectangle";
    }

    area(): number {
        return this.width * this.height;
    }
}

const shapes: Shape[] = [new Circle(2), new Rectangle(3, 4)];
// The loop does not branch on concrete type; each object supplies its own area.
for (const shape of shapes) {
    // The printed result shows whether the program can use dynamic dispatch without unsafe type
    // assumptions.
    console.log(`${shape.name()}: ${shape.area().toFixed(2)}`);
}

export {};
